### PR TITLE
Improved `RoomCoordinate` and `RoomXY`, new `RoomOffset` for optimized arithmetic

### PR DESCRIPTION
**Pull request by khoover**
Adds indexing impls for `RoomCoordinate` and a wrapper type that allows indexing nested arrays by `RoomXY`. Re-writes `LocalCostMatrix` and `LocalRoomTerrain` indexing in terms of the `RoomCoordinate` indexing impls.

Also adds a new `RoomOffset` type which can be used for better-optimized arithmetic, as well as being the output of coordinate-coordinate subtraction.

### CODE REVIEW

**Review by jciskey — APPROVED** [review]
LGTM, not seeing any major incompatibilities or issues. Even the documentation comments are solid.

**Comment by khoover** [comment]
The broken nightly checks are in unrelated files, about converting explicit lifetimes into implicit ones.

**Comment by khoover** [comment]
I think I can shave even more time off the distance transform in utils by using the `AsMut<XMajor<u8>> for LocalCostMatrix`.

File: src/local/room_coordinate.rs
Comment (khoover) [inline src/local/room_coordinate.rs:1]: There's actually some more stuff I'd like to add here, like `Deref<u8>` and a proper `RoomOffset` type for encoding coordinate-coordinate distances (would get to leverage it's restricted to `(-ROOM_SIZE, ROOM_SIZE)` when writing arithmetic).

**Comment by khoover** [comment]
What I'd want to do is replace the `Sub` impl for `RoomXY` with one that returns an `OffsetXY` or similar, instead of `i8` pairs.

**Pull request merged**